### PR TITLE
Make `jj squash <path>` not rewrite unmatched commits

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/rewrite.rs [hunk 1]
```diff
@@ -184,6 +184,13 @@ impl<'repo> CommitRewriter<'repo> {
         self.new_parents.retain(|parent| head_set.contains(parent));
     }
 
+    /// Records the old commit as abandoned with the new parents.
+    pub fn record_abandoned_commit(self) {
```
Comment (yuja) [inline lib/src/rewrite.rs:188]: nit: just call it `abandon()`?

I wouldn't expect `record_*()` is a consuming function, and it's similar to `.rebase()`.

File: lib/src/rewrite.rs [hunk 2]
```diff
@@ -184,6 +184,13 @@ impl<'repo> CommitRewriter<'repo> {
...
+        let old_commit_id =self.old_commit.id().clone();
```
Comment (yuja) [inline lib/src/rewrite.rs]: nit: `rustfmt` this commit

File: cli/src/commands/squash.rs [hunk 3]
```diff
@@ -213,6 +213,11 @@ from the source will be moved into the destination.
             diff_selector.select(&parent_tree, &source_tree, matcher, Some(&instructions))?;
         let selected_tree = tx.repo().store().get_root_tree(&selected_tree_id)?;
         let abandon = selected_tree.id() == source_tree.id();
+        if !abandon && selected_tree_id == parent_tree.id() {
```
Comment (yuja) [inline cli/src/commands/squash.rs:216]: nit: better to add comment why `!abandon`?

I thought an empty commit shouldn't be squashed, but apparently some tests rely on this behavior to move commit metadata back to the parent commit.

**Review by yuja — APPROVED** [review]

File: lib/src/rewrite.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/rewrite.rs]: Done

File: cli/src/commands/squash.rs (hunk 3 above)
Comment (martinvonz) [reply cli/src/commands/squash.rs:216]: Hopefully clarified now. Looks good?

Comment (yuja) [reply cli/src/commands/squash.rs:216]: Yes, thanks.

Btw, `jj squash -i` would previously error out in that situation. Just a comment, I don't think we'll want to add another special case. There's no way to distinguish "selected nothing" from "selected empty commit".

**Pull request merged**